dsgrid/config/dimensions.py

```diff
@@ -169,142 +238,145 @@ def check_name(cls, name):
+            if not os.path.isfile(filename):
+                raise ValueError(f"file {filename} does not exist")
+            if filename.startswith("s3://"):
+                raise ValueError("records must exist in the local filesystem, not on S3")
+            if not filename.endswith(".csv"):
+                raise ValueError(f"only CSV is supported: {filename}")
 
         return filename
 
-    @validator("file_hash")
-    def compute_file_hash(cls, file_hash, values):
-        if "filename" not in values:
-            return None
-        return file_hash or compute_file_hash(values["filename"])
+    @field_validator("file_hash")
+    @classmethod
+    def compute_file_hash(cls, file_hash, info: ValidationInfo):
+        if "filename" not in info.data:
+            return file_hash
 
-    @validator("records", always=True)
-    def add_records(cls, records, values):
+        if file_hash is None:
+            file_hash = compute_file_hash(info.data["filename"])
+        return file_hash
+
+    @field_validator("records")
+    @classmethod
+    def add_records(cls, records, info: ValidationInfo):
         """Add records from the file."""
-        if not {"filename", "cls", "name"}.issubset(values):
-            return None
-        dim_class = values["cls"]
+        dim_class = info.data.get("cls")
+        if "filename" not in info.data or dim_class is None:
+            return records
+
         if records:
             if isinstance(records[0], dict):
                 records = convert_record_dicts_to_classes(
                     records, dim_class, check_duplicates=["id"]
                 )
             return records
 
-        prereqs = ("name", "filename", "cls")
-        for req in prereqs:
-            if values.get(req) is None:
-                return records
-
-        filename = Path(values["filename"])
-        assert not str(filename).startswith("s3://"), "records must exist in the local filesystem"
-
-        if not filename.name.endswith(".csv"):
-            raise ValueError(f"only CSV is supported: {filename}")
-
-        with open(filename, encoding="utf") as f_in:
-            return convert_record_dicts_to_classes(
+        with open(info.data["filename"], encoding="utf-8") as f_in:
+            records = convert_record_dicts_to_classes(
                 csv.DictReader(f_in), dim_class, check_duplicates=["id"]
             )
+        return records
 
-    def dict(self, *args, **kwargs):
-        return super().dict(*args, **self._handle_kwargs(**kwargs))
-
-    def json(self, *args, **kwargs):
-        return super().json(*args, **self._handle_kwargs(**kwargs))
-
-    @staticmethod
-    def _handle_kwargs(**kwargs):
-        exclude = {"cls", "file", "filename"}
-        if "exclude" in kwargs and kwargs["exclude"] is not None:
-            kwargs["exclude"].union(exclude)
-        else:
-            kwargs["exclude"] = exclude
-        return kwargs
+    @field_serializer("cls", "filename")
```

Comment:
This is an example of how we can customize serialization of a field or exclude it.